middleware/rewrite/edns0.go

```diff
@@ -0,0 +1,132 @@
+// Package rewrite is middleware for rewriting requests internally to something different.
+package rewrite
+
+import (
+	"encoding/hex"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/miekg/dns"
+)
+
+// edns0LocalRule is a class rewrite rule.
+type edns0LocalRule struct {
+	action string
+	code   uint16
+	data   []byte
+	match  []byte // TODO: add match input for replace rule
+}
+
+type edns0NsidRule struct {
+	action string
+}
+
+func getEdns0Opt(r *dns.Msg) *dns.OPT {
```

Comment:
This doesn't only get, but also set the edns0 option if not found. Needs better name + some function doc.